bigframes/session/executor.py

```diff
@@ -161,20 +163,74 @@ def dry_run(self, array_value: bigframes.core.ArrayValue, ordered: bool = True):
         return results_iterator, query_job
 
     def peek(
-        self, array_value: bigframes.core.ArrayValue, n_rows: int
+        self,
+        array_value: bigframes.core.ArrayValue,
+        n_rows: int,
     ) -> tuple[bigquery.table.RowIterator, bigquery.QueryJob]:
-        """A 'peek' efficiently accesses a small number of rows in the dataframe."""
-        if not tree_properties.peekable(self._with_cached_executions(array_value.node)):
+        """
+        A 'peek' efficiently accesses a small number of rows in the dataframe.
+        """
+        plan = self._get_optimized_plan(array_value.node)
+        if not tree_properties.can_fast_peek(plan):
             warnings.warn("Peeking this value cannot be done efficiently.")
-        sql = self.compiler.compile_peek(
-            self._with_cached_executions(array_value.node), n_rows
-        )
+
+        sql = self.compiler.compile_peek(plan, n_rows)
 
         # TODO(swast): plumb through the api_name of the user-facing api that
         # caused this query.
-        return self._run_execute_query(
-            sql=sql,
-        )
+        return self._run_execute_query(sql=sql)
+
+    def head(
+        self, array_value: bigframes.core.ArrayValue, n_rows: int
+    ) -> tuple[bigquery.table.RowIterator, bigquery.QueryJob]:
+        """
+        A 'peek' efficiently accesses a small number of rows in the dataframe.
+        """
+        maybe_row_count = self._local_get_row_count(array_value)
+        if (maybe_row_count is not None) and (maybe_row_count <= n_rows):
+            return self.execute(array_value, ordered=True)
+
+        if not self.strictly_ordered and not array_value.node.explicitly_ordered:
+            # No user-provided ordering, so just get any N rows, its faster!
+            return self.peek(array_value, n_rows)
+
+        plan = self._get_optimized_plan(array_value.node)
+        if not tree_properties.can_fast_head(plan):
+            # If can't get head fast, we are going to need to execute the whole query
+            # Will want to do this in a way such that the result is reusable, but the first
+            # N values can be easily extracted. This means clustering on the order key.
+
+            # This may be too much, might be sufficient to cluster on ordering key (if <= 4 parts)
+            self._cache_with_offsets(array_value)
+            # Get a new optimized plan after caching
+            plan = self._get_optimized_plan(array_value.node)
+            assert tree_properties.can_fast_head(plan)
+
+        head_plan = generate_head_plan(plan, n_rows)
+        sql = self.compiler.compile_ordered(head_plan)
+
+        # TODO(swast): plumb through the api_name of the user-facing api that
+        # caused this query.
+        return self._run_execute_query(sql=sql)
+
+    def get_row_count(self, array_value: bigframes.core.ArrayValue) -> int:
+        # optimized plan less likely to have count-destroying operators like filter or join
```

Comment:
What's this comment mean? Do you mean that if the count is not None, then we have an optimized plan?

Reply:
removed comment. essential idea is to use the row_num metadata from cached execution where possible